### PR TITLE
Incremental Training inside Rasa Open Source

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- Closes https://github.com/RasaHQ/rasa/issues/6971
- Adds the ability to load NLU components and Core policies in finetune mode which means they are initialized with a previously trained model and can be further trained on a mix of old and new training data.

Individual PRs:
- Changes to CLI - https://github.com/RasaHQ/rasa/issues/6971
- Applying constraints for finetuning - https://github.com/RasaHQ/rasa/pull/7427
   - Configuration of base model is exactly the same as configuration for finetuning except the number of epochs.
   - No new labels(intents, actions, entities and intent_response_keys) are added to the training data. New training examples for each of these can still be added.
   - Base model is trained with `MINIMUM_COMPATIBLE_VERSION` of currently installed Rasa Open Source version.
- Load NLU(https://github.com/RasaHQ/rasa/pull/7456) and Core(https://github.com/RasaHQ/rasa/pull/7458) pipelines in finetune mode.
- Loading NLU components and Core policies in finetune mode - https://github.com/RasaHQ/rasa/pull/7419 This PR includes running model regression tests to verify that performance of current models are not hurt by the extra featurization we do.
- Docs (https://github.com/RasaHQ/rasa/pull/7469)

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: docs/docs/command-line-interface.mdx [hunk 1]
```diff
@@ -85,6 +85,48 @@ The following arguments can be used to configure the training process:
...
+
+To be able to fine tune a model, the following conditions must be met:
+
+1. The configuration supplied should be exactly the same as the
+configuration used to train the model which is being finetuned.
+The only parameter that you can change is `epochs` for the individual machine learning components and policies.
+
+2. The domain supplied should be exactly the same as the domain used to train the model initially. 
+That means no new intents, actions, entities or slots can be added to the domain.
+Only new training examples can be added.
```
Comment (dakshvar22) [inline docs/docs/command-line-interface.mdx:117]:
```suggestion
2. The set of labels(intents, actions and entities or slots) for which the base model is trained should be exactly the same as the ones present in the training data for finetuning. This means that you cannot add new intent, action, entity or slot labels during incremental training. You can still add new training examples for each of the existing labels. If you have added new labels to the training data, it is advisable to train the pipeline from scratch.
```

Comment (dakshvar22) [reply docs/docs/command-line-interface.mdx:117]: @wochinge @joejuzl Does this accurately reflect what we discussed on the call?

Comment (joejuzl) [reply docs/docs/command-line-interface.mdx:117]: Yes although maybe `advisable`->`required`?

Comment (dakshvar22) [reply docs/docs/command-line-interface.mdx:117]: Good point. Changed it.

File: changelog/7498.feature.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+Incremental training of models in a pipeline is now supported.
+
+If you have added new NLU training examples or new stories/rules for  
+dialogue manager, you don't need to train the pipeline from scratch.
+Instead, you can initialize the pipeline with a previously trained model
+and continue finetuning the model on the complete dataset consisting of
+new training examples. To do so, use `rasa train --finetune`. For more
+detailed explanation of the command, check out the docs on [incremental  
+training](./command-line-interface.mdx#incremental-training)
```
Comment (wochinge) [inline changelog/7498.feature.md:9]:
```suggestion
training](./command-line-interface.mdx#incremental-training).
```

File: changelog/7498.feature.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+Incremental training of models in a pipeline is now supported.
```
Comment (wochinge) [inline changelog/7498.feature.md:1]: don't we an issue for incremental training? In that case we should use the issue number in the changelog name.

**Comment by dakshvar22** [comment]
Okay, so this turned out to be a large PR as well but all the smaller PRs were peer reviewed between @joejuzl @wochinge and @dakshvar22. Additionally, PR for changes to the ML components(https://github.com/RasaHQ/rasa/pull/7419) was reviewed by Tanja as well.

@alwx Can you please review the following files - `rasa/cli/...`, `rasa/core/agent.py`, `rasa/core/train.py`, `rasa/model.py`, `rasa/nlu/model.py`, `rasa/nlu/train.py`, `rasa/telemetry.py`, `rasa/train.py`, `tests/cli/test_rasa_train.py`, `tests/conftest.py`, `tests/core/conftest.py`, `tests/core/test_agent.py`, `tests/core/test_ensemble.py`, `tests/test_train.py`? These are the engineering bits of the PR.

File: changelog/6971.feature.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+If you have added new NLU training examples or new stories/rules for  
+dialogue manager, you don't need to train the pipeline from scratch.
+Instead, you can initialize the pipeline with a previously trained model
+and continue finetuning the model on the complete dataset consisting of
+new training examples. To do so, use `rasa train --finetune`. For more
+detailed explanation of the command, check out the docs on [incremental  
+training](./command-line-interface.mdx#incremental-training).
+
+Added a configuration parameter `additional_vocabulary_size` to  
+[`CountVectorsFeaturizer`](./components.mdx#countvectorsfeaturizer)
```
Comment (Ghostvv) [inline changelog/6971.feature.md:12]: are you sure it is `./components.mdx` and not `components.mdx`? Asking because I need to reference docs in my PR as well

File: docs/docs/components.mdx [hunk 5]
```diff
@@ -836,6 +836,28 @@ Note: The `feature-dimension` for sequence and sentence features does not have t
...
+  parameter while training the base model from scratch:
+
+  ```yaml-rasa {3}
+  pipeline:
+  - name: RegexFeaturizer
+    number_additional_patterns: 10
+  ```
+
+  If not configured by the user, the component will account for a minimum of 10 additional
+  patterns and a maximum of twice the number of patterns currently
```
Comment (Ghostvv) [inline docs/docs/components.mdx:854]: what does it mean? it is not clear to me what is the default `number_additional_patterns`

File: docs/docs/components.mdx [hunk 6]
```diff
@@ -960,58 +982,91 @@ Note: The `feature-dimension` for sequence and sentence features does not have t
...
+      response: 1000
+      action_text: 1000
+  ```
+
+  As in the above example, you can define additional vocabulary size for each of
+  `text` (user messages), `response` (bot responses used by `ResponseSelector`) and
+  `action_text` (bot responses not used by `ResponseSelector`). If you are building a shared
+  vocabulary (`use_shared_vocab=True`), you only need to define a value for the `text` attribute.
+  If any of the attributes is not configured by the user, the component will
+  account for a minimum of 1000 additional vocabulary slots and a maximum of half the current vocabulary size.
```
Comment (Ghostvv) [inline docs/docs/components.mdx:1007]: same here, it is not clear what is the default number

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 7]
```diff
@@ -143,8 +149,20 @@ def _load_OOV_params(self) -> None:
             if self.OOV_words:
                 self.OOV_words = [w.lower() for w in self.OOV_words]
 
+        # Additional vocabulary size to be kept reserved
+        self.additional_vocabulary_size = self.component_config[
+            "additional_vocabulary_size"
+        ]
+        if isinstance(self.additional_vocabulary_size, int):
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:156]: do we allow both `int` and `dict`?

Comment (Ghostvv) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:156]: I suggest to always have `dict`

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 8]
```diff
@@ -341,72 +361,257 @@ def _convert_attribute_tokens_to_texts(
...
+            return 0
+
+        configured_additional_size = self.additional_vocabulary_size.get(attribute)
+        if configured_additional_size is not None:
+            return configured_additional_size
+
+        # If the user hasn't defined additional vocabulary size,
+        # then we increase it by 1000 minimum. If the current
+        # vocabulary size is greater than 2000, we take half of
+        # that number as additional vocabulary size.
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:440]: do you think there is a need for such complexity?

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 9]
```diff
@@ -341,72 +361,257 @@ def _convert_attribute_tokens_to_texts(
...
+                    "max_df": self.max_df,
+                    "min_df": self.min_df,
+                    "max_features": self.max_features,
+                    "analyzer": self.analyzer,
+                }
+            )
+            self._fit_vectorizer_from_scratch(TEXT, combined_cleaned_texts)
+        else:
+            self._fit_loaded_vectorizer(TEXT, combined_cleaned_texts)
+        self._log_vocabulary_stats(TEXT)
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:523]: could you please add comment why `TEXT` is hardcoded here

File: rasa/utils/tensorflow/models.py [hunk 10]
```diff
@@ -341,8 +360,10 @@ def load(
             evaluate_on_num_examples=0,
             batch_strategy=SEQUENCE,
             silent=True,  # don't confuse users with training output
-            loading=True,  # don't use tensorboard while loading
-            eager=True,  # no need to build tf graph, eager is faster here
+            loading=True,  # don't use tensorboard when doing a dummy fit run
+            eager=False
+            if finetune_mode
+            else True,  # load in eager mode only for prediction phase
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/models.py:366]: not sure if `black` will complain, but parenthesis make it easier to spot that this is all related to `eager=`
```suggestion
            eager=(False
                if finetune_mode
                else True
            ),  # load in eager mode only for prediction phase
```

File: tests/core/policies/test_rule_policy.py [hunk 11]
```diff
@@ -656,6 +656,48 @@ def test_all_policy_attributes_are_persisted(tmpdir: Path):
     assert persisted_policy._enable_fallback_prediction == enable_fallback_prediction
 
 
+async def test_rule_policy_finetune(
```
Comment (Ghostvv) [inline tests/core/policies/test_rule_policy.py:659]: could you please add a test that contradiction is raised if you add contradicting rule in fine-tune

**Review by Ghostvv — COMMENTED** [review]
looks good from my side. left a couple of comments

File: changelog/6971.feature.md (hunk 4 above)
Comment (dakshvar22) [reply changelog/6971.feature.md:12]: Both should work since the `changelog.mdx` is in the same folder as `components.mdx`.

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py (hunk 8 above)
Comment (dakshvar22) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:440]: From experiments, we noticed that using more than 0.5 times(0.75 and above) the current vocabulary size hurt the performance of training a model from scratch. The minimum of 1000 was kept so that users don't need to finetune frequently.

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py (hunk 7 above)
Comment (dakshvar22) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:156]: Ahh good that you pointed it out, I had meant to remove this piece of code. Agreed, we should stick to `dict`.

File: docs/docs/components.mdx (hunk 5 above)
Comment (dakshvar22) [reply docs/docs/components.mdx:854]: Rephrased. Is it clearer now?

File: docs/docs/components.mdx (hunk 6 above)
Comment (dakshvar22) [reply docs/docs/components.mdx:1007]: Rephrased. Is it clearer now?

File: tests/core/policies/test_rule_policy.py (hunk 11 above)
Comment (dakshvar22) [reply tests/core/policies/test_rule_policy.py:659]: Added

File: rasa/cli/arguments/train.py [hunk 12]
```diff
@@ -21,52 +24,63 @@ def set_train_arguments(parser: argparse.ArgumentParser):
...
+    _add_model_name_param(parser)
     add_persist_nlu_data_param(parser)
+    add_finetune_params(parser)
 
 
-def add_force_param(parser: Union[argparse.ArgumentParser, argparse._ActionsContainer]):
+def add_force_param(
+    parser: Union[argparse.ArgumentParser, argparse._ActionsContainer]
+) -> None:
+    """Specifies if the model should be trained from scratch."""
```
Comment (alwx) [inline rasa/cli/arguments/train.py:72]: `Args` section is missing in the docstring but that's probably something CI will complain about.

**Review by Ghostvv — APPROVED** [review]
the files that I was assigned to look good

File: rasa/nlu/featurizers/sparse_featurizer/regex_featurizer.py [hunk 13]
```diff
@@ -160,23 +291,58 @@ def load(
...
-
+        """Loads a previously trained component.
+
+        Args:
+            meta: Configuration of trained component.
+            model_dir: Path where trained pipeline is stored.
+            model_metadata: Metadata for the trained pipeline.
+            cached_component: Previously cached component(if any).
+            should_finetune: Indicates whether to load the component for further finetuning.
+            **kwargs:
```
Comment (alwx) [inline rasa/nlu/featurizers/sparse_featurizer/regex_featurizer.py:305]: Description of the argument is missing

File: tests/cli/test_rasa_train.py [hunk 14]
```diff
@@ -419,7 +422,8 @@ def test_train_core_help(run: Callable[..., RunResult]):
                        [--augmentation AUGMENTATION] [--debug-plots] [--force]
                        [--fixed-model-name FIXED_MODEL_NAME]
                        [--percentages [PERCENTAGES [PERCENTAGES ...]]]
-                       [--runs RUNS]"""
+                       [--runs RUNS] [--finetune [FINETUNE]]
+                       [--epoch-fraction EPOCH_FRACTION]"""
```
Comment (alwx) [inline tests/cli/test_rasa_train.py:426]: There are no tests for the command itself, right? I think it makes sense to add them.
(something with `run_in_simple_project("train", "--finetune"`)

**Review by alwx — CHANGES_REQUESTED** [review]

File: rasa/cli/arguments/train.py (hunk 12 above)
Comment (wochinge) [reply rasa/cli/arguments/train.py:72]: It's missing on purpose because I don't see any value in having full docstrings for these helper functions. The CI actually allows one-line docstrings 😁 What do you think?

File: tests/cli/test_rasa_train.py (hunk 14 above)
Comment (joejuzl) [reply tests/cli/test_rasa_train.py:438]: I've done a basic cli test with the command (which just tests the arg is actually passed down). And I've changed `test_model_finetuning` to not mock the train methods and actually run the second training.

**Comment by dakshvar22** [comment]
A final run of model regression tests before merge is running [here](https://github.com/RasaHQ/rasa/pull/7544)

**Review by alwx — APPROVED** [review]
